### DISCUSSION TITLE
2FA with radius and tpm

### DISCUSSION THREAD

**Discussion by bmoe3rd** [strongswan/strongswan#271] [Q&A]
Hello everyone,

currently i try to understand how to best do 2FA with strongSwan and FreeRadius.

As i must use a tpm for private key storage, i would like to use strongSwan for authenticating
users with certificates.
Radius would then do the second round of authentication with passwords.

1. Is there a strong connection between password and certificate ?
    ( it should be impossible to login with another users password )
2. Does the radius server create an encrypted tunnel to the client ?
    ( if not, how is it done and what are my options ? )
3. Is it possible to connect strongSwan to the radius server without using certificates?
    ( strongSwan and FreeRadius are on the same machine and i must secure all private keys in hardware )
4. And last but not least, are there examples with swanctl? ;-)

EDIT:
to 4: I found the swanctl examples under "swanctl" T_T. I searched in ikev2 until now.
to 2: is it right, that eap-radius configures that eap is handled by radius only?
        So wich eap is used is configured with FreeRadius entirely?
        How does Carol know about wich eap is used then?

thank you for your help in advance,
bmoe3rd

**Comment by Thermi** [comment]
Hi,

1) No, not implemented in strongSwan. I do not know if that is possible to implement using the RADIUS server.
2) No, EAP itself is not encrypted. It is just a framework. If anything is effectively encrypted depends on if you use an EAP method that does that. The connection between IKE daemon and initiator is always encrypted, but the question is if it is authenticated. Ususally, it's authenticated by the client by authenticating the server's certificate against the configured identity for the remote peer. The RADIUS protocol effectively is only obfuscated by the shared key between RADIUS server and client.
3) Yes, there is no cryptographic binding between the RADIUS server and client besides the shared RADIUS secret.
4) Just look at any eap-radius example.
5) Only for the conn you use eap-radius in.

6) It's negotiated using EAP NAK and ACKs between RADIUS server and RADIUS client, like with eap-dynamic.

**Comment by strongX509** [comment]
- EAP-TLS establishes an encrypted end-to-end connection between the client and the RADIUS server, with both the client and server authenticating themselvers with  a certificate.
- But if each client has a certificate stored in the TPM you won't need a RADIUS server because the VPN gateway could verify the client certificates based on the CA certifcate.
- EAP-TTLS or the Microsoft EAP-PEAP junk protocol establishes an encrypted end-to-end connection between the client and the RADIUS server, with the server authenticating itself with a certificate and the client authenticating itself with a user password through the secure EAP connection using  either EAP-MD5 or EAP-MSChapV2.
- In principle you could do a multiple authentication with a first round based on client certificates verified by the VPN gateway and a second round with a secure EAP-TTLS connection to a RADIUS server based on user passwords.
